util/include/Handoff.hpp

```diff
@@ -12,104 +12,272 @@
  */
 class Handoff {
-  typedef std::lock_guard<std::mutex> guard;
-  typedef std::function<void()> func_type;
+ protected:
+  using guard = std::lock_guard<std::mutex>;
+  using func_t = std::function<void()>;
+  struct HandoffElement;
 
  public:
   Handoff(std::uint16_t replicaId, const std::string&& name = "")
-      : logger_{logging::getLogger("concord.util.handoff")} {
-    if (!name.empty()) {
-      name_ = std::move(name);
-    }
-    log_prefix_ = name_ + ": ";
+      : tasks_(name + "_PriorityQueue"),
+        name_(std::move(name)),
+        log_prefix_(name_ + ": "),
+        logger_{logging::getLogger("concord.util.handoff")},
+        element_id_counter_{0} {
+    // keep for debugging
+    SET_LOG_LEVEL(log4cplus::TRACE_LOG_LEVEL);
+
     thread_ = std::thread([this, replicaId] {
       try {
         MDC_PUT(MDC_REPLICA_ID_KEY, std::to_string(replicaId));
         MDC_PUT(MDC_THREAD_KEY, "handoff");
-        for (;;) pop()();
+        DEBUG_PRINT(logger_, log_prefix_ << "consumer: entering loop...");
+        for (;;) {
+          HandoffElement element = pop();
+          // keep for debug
+          DEBUG_PRINT(logger_,
+                      std::boolalpha << log_prefix_ << "consumer: before calling id=" << element.id
+                                     << " blocked_call:" << (element.blocked_call));
+          element.func();
+          DEBUG_PRINT(logger_, log_prefix_ << "consumer: after calling id=" << element.id);
+
+          if (element.blocked_call) {
+            std::unique_lock<std::mutex> ul(tasks_lock_);
+            element.completion_flag = true;
+            element.completion_cond.notify_one();
+          }
+        }
       } catch (ThreadCanceledException& e) {
-        LOG_INFO(logger_, log_prefix_ << "thread cancelled " << std::this_thread::get_id());
+        LOG_INFO(logger_, log_prefix_ << "thread cancelled:" << KVLOG(thread_id_.load()));
       } catch (const std::exception& e) {
         LOG_FATAL(logger_, log_prefix_ << "exception: " << e.what());
         std::terminate();
       }
     });
+    thread_id_.store(thread_.get_id());
+    LOG_INFO(logger_, log_prefix_ << "started handoff consumer thread:" << KVLOG(thread_id_.load()));
   }
 
   ~Handoff() { stop(); }
 
   void stop() {
-    if (stopped_.exchange(true)) return;
-    queue_cond_.notify_one();
-    auto tid = thread_.get_id();
+    if (stopped_.exchange(true)) {
+      return;
+    }
+    tasks_cond_.notify_one();
     thread_.join();
-    LOG_INFO(logger_, log_prefix_ << "thread joined " << tid);
+    LOG_INFO(logger_, log_prefix_ << "thread joined:" << KVLOG(thread_id_.load()));
   }
 
-  void push(func_type f) {
+  // By default, calls are blocked (synchronized)
+  void push(func_t&& f, bool blocked_call = true) {
+    std::condition_variable completion_cond;
+    bool completion_flag{false};
+
+    auto thread_id = thread_id_.load();
+    if (blocked_call && (thread_id == std::this_thread::get_id())) {
```

Comment:
It actually happened in the run few times, before I refactored the code further in multile ways.
In general, any time you decide to call ST interface in ST main thread context, the consumer becomes also a producer. In some cases it is avoiable by calling the impl function, but in some cases someone who is not familiar with the code mayb call the API, which is not efficient but legit.
Before I made isCollectingState an atomic API, this was one example, but there were few more in the callbacks at the end of cycle, which later transformed to be execute by a worker thread.
Anyhow even if there is nothing today, its still a valid mechanism to avoid a sure deadlock.